src/value.cpp

```diff
@@ -906,51 +907,30 @@ namespace Exiv2 {
 
     int DateValue::read(const byte* buf, long len, ByteOrder /*byteOrder*/)
     {
-        // Hard coded to read Iptc style dates
-        if (len != 8) {
-#ifndef SUPPRESS_WARNINGS
-            EXV_WARNING << Error(kerUnsupportedDateFormat) << "\n";
-#endif
-            return 1;
-        }
-        // Make the buffer a 0 terminated C-string for sscanf
-        char b[] = { 0, 0, 0, 0, 0, 0, 0, 0, 0 };
-        std::memcpy(b, reinterpret_cast<const char*>(buf), 8);
-        int scanned = sscanf(b, "%4d%2d%2d",
-                             &date_.year, &date_.month, &date_.day);
-        if (   scanned != 3
-            || date_.year < 0
-            || date_.month < 1 || date_.month > 12
-            || date_.day < 1 || date_.day > 31) {
-#ifndef SUPPRESS_WARNINGS
-            EXV_WARNING << Error(kerUnsupportedDateFormat) << "\n";
-#endif
-            return 1;
-        }
-        return 0;
+        const std::string str(reinterpret_cast<const char*>(buf), len);
+        return read(str);
     }
 
     int DateValue::read(const std::string& buf)
     {
-        // Hard coded to read Iptc style dates
-        if (buf.length() < 8) {
-#ifndef SUPPRESS_WARNINGS
-            EXV_WARNING << Error(kerUnsupportedDateFormat) << "\n";
-#endif
-            return 1;
+        // ISO 8601 date formats:
+        // https://web.archive.org/web/20171020084445/https://www.loc.gov/standards/datetime/ISO_DIS%208601-1.pdf
+        static const std::regex reExtended(R"(^(\d{4})-(0[1-9]|1[0-2])-(0[1-9]|[12][0-9]|3[01]))");
+        static const std::regex reBasic(R"(^(\d{4})(0[1-9]|1[0-2])(0[1-9]|[12][0-9]|3[01]))");
```

Comment:
As discussed [here](https://github.com/Exiv2/exiv2/pull/1955), there is a possibility to support years before BC in ISO 8601 (unless IPTC expressly prohibits these), so an optional [+-] prefix for the year could be included as well?

Reply:
According to the ISO document, only values from 0000 to 9999 should be considered:

> calendar year is, unless specified otherwise, represented by four digits. Calendar years are
numbered in ascending order according to the Gregorian calendar by values in the range [0000]
to [9999]. Values in the range [0000] through [1582] shall only be used by mutual agreement of
the partners in information interchange.

I actually took these regular expressions from the book "Regular Expressions Cookbook":
https://learning.oreilly.com/library/view/regular-expressions-cookbook/9781449327453/ch04s04.html

Reply:
The standard does allow for negative years in the "expanded form" (to be agreed between exchanging parties), [as mentioned on Wikipedia](https://en.wikipedia.org/wiki/ISO_8601#Years). Just don't know if that's applicable here or not, not that familiar w/ IPTC...

Reply:
From your wikipedia link:

> SO 8601 prescribes, as a minimum, a four-digit year [YYYY] to avoid the year 2000 problem. It therefore represents years from 0000 to 9999, year 0000 being equal to 1 BC and all others AD. However, years before 1583 are not automatically allowed by the standard. Instead "values in the range [0000] through [1582] shall only be used by mutual agreement of the partners in information interchange."[19]. To represent years before 0000 or after 9999, the standard also permits the expansion of the year representation but only by prior agreement between the sender and the receiver.[20] 

IMHO, I think anything below 1900 is not realistic for representing Creation/Modification dates in digital images. I would not add the optional [+|-] unless more of you guys think it would cover some valid cases.

Reply:
Scans of 19th century photographs can have creation dates prior to 1900.  (: